### PR TITLE
Bicolor gradient bug fix and added optional PNG export to save function

### PR DESCRIPTION
**Pull request by elizabethmcd**
This PR fixes/updates:
- Bug where bicolor gradient positions were incorrect #34 
- Adds option to `save_arcadia_plot` function to export as PNG #35, where by default it will only save PDF but can optionally save just PNG, or both PDF and PNG. I had to convert pixels to inches for the resolution to look correct, even at that the resolution isn't as sharp as PDF exports
- Updates documentation 

I will cut a new version after this since this has a bug fix in it.

### CODE REVIEW

File: man/figures/README-heatmap_example-1.png
Comment (mezarque) [inline man/figures/README-heatmap_example-1.png:1]: While the weird banding seems to be fixed, the relative position of colors still seems to be wrong here; the amount of light color is much more exaggerated than it should be. Below is an example of the gradient implemented in Illustrator, placed above a screenshot of the bicolor gradient.

Screenshot 2024-07-31 at 1 44 58 PM

File: R/arcadia_palette.R [hunk 1]
```diff
@@ -216,12 +216,16 @@ BLUES <- list(colors = c(LAPIS, AEGEAN, ZEPHYR), positions = c(0, 0.254, 1.0))
...
+  # rescale positions to fit within [0, 0.5] and [0.5, 1]
+  rescaled_grad1_positions <- grad1$positions * 0.5
+  rescaled_grad2_positions <- grad2$positions * 0.5 + 0.5
+
+  # combine colors and the rescaled positions
   colors <- c(grad1$colors, rev(grad2$colors))
   # Rescale the positions to fit the combined gradient
-  positions <- c(grad1$positions, rev(grad2$positions) + (1 - max(grad2$positions)))
+  positions <- c(rescaled_grad1_positions, rescaled_grad2_positions)
   list(colors = colors, positions = scales::rescale(positions))
```
Comment (mezarque) [inline R/arcadia_palette.R:229]: Do we still need `scales::rescale` here?

File: R/save_arcadia_plot.R [hunk 2]
```diff
@@ -1,57 +1,86 @@
 #' Save Arcadia Plot
 #'
-#' @param plot ggplot object to be saved
+#' @param plot ggplot object to be saved as both a PDF and PNG
 #' @param filepath The path where the plot will be saved
```
Comment (mezarque) [inline R/save_arcadia_plot.R:4]: Should this be amended to say "without file extension"?

File: man/figures/README-heatmap_example-1.png
Comment (mezarque) [reply man/figures/README-heatmap_example-1.png:1]: Not sure why this is happening, because the position values in the output of the README seem to be correct... but the final plot is not correct.

**Review by mezarque — CHANGES_REQUESTED** [review]
Added a few thoughts! Seems like the bicolor gradient still needs a little fussing with to get it to work properly.

**Review by reillyowencooper — COMMENTED** [review]
Everything looks like it works in the save function! The way you changed the panel sizing is really elegant compared to the prior version. I added one inline question about the default arg parameters for `panel_size`, since I'm not sure how the logic works with a list getting passed in and the `size <- PANEL_OPTIONS[[panel_size]]` line calling a value from that. I might be interpreting that incorrectly though. I didn't comment on the gradients since I don't understand the math with those.

File: R/arcadia_palette.R (hunk 1 above)
Comment (elizabethmcd) [reply R/arcadia_palette.R:229]: Oh yeah probably not, hopefully that isn't what was causing the issues still..

File: R/save_arcadia_plot.R [hunk 3]
```diff
@@ -1,57 +1,86 @@
...
-#' save_arcadia_plot("plot_full_square.pdf", p, "full_square")
-#' save_arcadia_plot("plot_float_square.pdf", p, "float_square")
+#' save_arcadia_plot("plot_full_wide", p, "full_wide")
+#' save_arcadia_plot("/path/to/figs/folder/plot_float_wide", p, "float_wide")
+#' save_arcadia_plot("plot_half_square", p, "half_square"
+#' save_arcadia_plot("plot_full_square", p, "full_square")
+#' save_arcadia_plot("plot_float_square", p, "float_square")
 #' }
-save_arcadia_plot <- function(filepath, plot, panel_size = c("full_wide", "float_wide", "half_square", "full_square", "float_square")) {
+save_arcadia_plot <- function(filepath, plot, panel_size = names(PANEL_OPTIONS), transparent_bg = TRUE, formats = c("pdf")) {
```
[the remark this answers is not in the document]
Comment (elizabethmcd) [reply R/save_arcadia_plot.R:21]: good point, I didn't set a default

**Comment by elizabethmcd** [comment]
@reillyowencooper I incorporated your suggestions. If you would like to take a stab at trying to fix the bicolor gradient positions problem I would greatly appreciate it. As @mezarque mentions above, I'm pretty sure the actual positions are correct but somehow it's not carrying through correctly to the heatmap.

File: R/save_arcadia_plot.R [hunk 4]
```diff
@@ -1,57 +1,93 @@
...
-#' save_arcadia_plot("plot_full_wide.pdf", p, "full_wide")
-#' save_arcadia_plot("plot_float_wide.pdf", p, "float_wide")
-#' save_arcadia_plot("plot_half_square.pdf", p, "half_square"
-#' save_arcadia_plot("plot_full_square.pdf", p, "full_square")
-#' save_arcadia_plot("plot_float_square.pdf", p, "float_square")
+#' save_arcadia_plot("plot_full_wide", p, "full_wide")
+#' save_arcadia_plot("/path/to/figs/folder/plot_float_wide", p, "float_wide", formats = c("pdf", "png"))
+#' save_arcadia_plot("plot_half_square", p, "half_square", formats = "png")
+#' save_arcadia_plot("plot_full_square", p, "full_square")
+#' save_arcadia_plot("plot_float_square", p, "float_square")
```
Comment (taylorreiter) [inline R/save_arcadia_plot.R:19]: can you name all of your arguments here? there are enough now where it is hard to follow

File: man/figures/README-heatmap_example-1.png
Comment (reillyowencooper) [reply man/figures/README-heatmap_example-1.png:1]: > While the weird banding seems to be fixed, the relative position of colors still seems to be wrong here; the amount of light color is much more exaggerated than it should be. Below is an example of the gradient implemented in Illustrator, placed above a screenshot of the bicolor gradient.
> 
> Screenshot 2024-07-31 at 1 44 58 PM

Dropping my thought here since I don't have time today to get to this, but I think the problem is that `geom_tile()` doesn't have a great way to interpret colors with custom positions. It's probably just making the values equidistant across the range, which is why the white values are exaggerated. If this is the case, using `scale_fill_gradientn(colors = [bicolor_gradient]$colors,
                       values = [bicolor_gradient]$positions,
                       limits = range(data$value),
                       guide = "colorbar")`
                       
might resolve the issue

File: R/save_arcadia_plot.R (hunk 2 above)
Comment (elizabethmcd) [reply R/save_arcadia_plot.R]: addressed

File: R/save_arcadia_plot.R [hunk 5]
```diff
@@ -1,57 +1,86 @@
...
   panel_size <- tryCatch(match.arg(panel_size), error = function(e) NA)
 
   if (is.na(panel_size)) {
-    stop("Invalid panel_size. Please choose from: 'full_wide', 'float_wide', 'half_square', 'full_square', 'float_square'.")
+    stop(paste("Invalid panel_size. Please choose from:", paste(names(PANEL_OPTIONS), collapse = ", "), "."))
   }
 
+  # ensure formats argument is valid
+  valid_formats <- c("pdf", "png")
+  formats <- match.arg(formats, choices = valid_formats, several.ok = TRUE)
```
[the remark this answers is not in the document]
Comment (elizabethmcd) [reply R/save_arcadia_plot.R]: addressed

File: R/save_arcadia_plot.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (elizabethmcd) [reply R/save_arcadia_plot.R]: addressed

File: R/save_arcadia_plot.R (hunk 4 above)
Comment (elizabethmcd) [reply R/save_arcadia_plot.R:19]: addressed

File: R/save_arcadia_plot.R [hunk 6]
```diff
@@ -1,57 +1,93 @@
 #' Save Arcadia Plot
 #'
 #' @param plot ggplot object to be saved
-#' @param filepath The path where the plot will be saved
-#' @param panel_size The size of the panel, one of "full_wide", "float_wide", "half_square", "full_square", "float_square"
+#' @param filepath The path where the plot will be saved without a file extension
```
[the remark this answers is not in the document]
Comment (elizabethmcd) [reply R/save_arcadia_plot.R:4]: addressed

**Comment by elizabethmcd** [comment]
@reillyowencooper @mezarque 
I had to play around with this for a while, but it's not a problem specific to `geom_tile`, it's that R doesn't like that the two white-ish colors are both at 0.5, so I had to put one at 0.5 and one at 0.55 for example, and it gets closer to what it's supposed to be. The other colors positions are the same as before, but because this was fixed through manual fiddling I just made these palettes also named vectors with those positions instead of a hard calculation since it doesn't seem to behave the same way as in python.

File: R/arcadia_palette.R [hunk 7]
```diff
@@ -216,19 +216,11 @@ BLUES <- list(colors = c(LAPIS, AEGEAN, ZEPHYR), positions = c(0, 0.254, 1.0))
...
 
-# Bicolor gradients
-# combine the gradients and inherit positions
-combine_gradients <- function(grad1, grad2) {
-  colors <- c(grad1$colors, rev(grad2$colors))
-  # Rescale the positions to fit the combined gradient
-  positions <- c(grad1$positions, rev(grad2$positions) + (1 - max(grad2$positions)))
-  list(colors = colors, positions = scales::rescale(positions))
-}
+ORANGE_SAGE <- list(colors = c(TERRACOTTA, TANGERINE, DAWN, LICHEN, SAGE, ASPARAGUS), positions = c(0, 0.3805, 0.5, 0.55, 0.8205, 1))
```
Comment (mezarque) [inline R/arcadia_palette.R:219]: This solution seems fine for now, but I would make sure that the lightest colors are symmetric about 0.5 to minimize any skew (assuming these particular values don't result in the exact same issue).

```suggestion
ORANGE_SAGE <- list(colors = c(TERRACOTTA, TANGERINE, DAWN, LICHEN, SAGE, ASPARAGUS), positions = c(0, 0.3805, 0.49, 0.51, 0.8205, 1))
```

File: R/arcadia_palette.R [hunk 8]
```diff
@@ -216,19 +216,11 @@ BLUES <- list(colors = c(LAPIS, AEGEAN, ZEPHYR), positions = c(0, 0.254, 1.0))
...
+
+RED_BLUE <- list(colors = c(CINNABAR, DRAGON, BLUSH, ZEPHYR, AEGEAN, LAPIS), positions = c(0, 0.106, 0.5, 0.55, 0.627, 1))
```
Comment (mezarque) [inline R/arcadia_palette.R:221]:
```suggestion
RED_BLUE <- list(colors = c(CINNABAR, DRAGON, BLUSH, ZEPHYR, AEGEAN, LAPIS), positions = c(0, 0.106, 0.49, 0.51, 0.627, 1))
```

File: R/arcadia_palette.R [hunk 9]
```diff
@@ -216,19 +216,11 @@ BLUES <- list(colors = c(LAPIS, AEGEAN, ZEPHYR), positions = c(0, 0.254, 1.0))
...
 
-# composite gradients
-ORANGE_SAGE <- combine_gradients(ORANGES, SAGES)
-RED_BLUE <- combine_gradients(REDS, BLUES)
-PURPLE_GREEN <- combine_gradients(PURPLES, GREENS)
+PURPLE_GREEN <- list(colors = c(LILAC, ASTER, GHOST, LICHEN, LIME, FERN), positions = c(0, .072, 0.5, 0.55, 0.811, 1))
```
Comment (mezarque) [inline R/arcadia_palette.R:223]:
```suggestion
PURPLE_GREEN <- list(colors = c(LILAC, ASTER, GHOST, LICHEN, LIME, FERN), positions = c(0, .072, 0.49, 0.51, 0.811, 1))
```

**Review by mezarque — CHANGES_REQUESTED** [review]
Could you check whether intermediate values of `0.49` and `0.51` will work? Otherwise the gradient will be skewed off towards one side, which could lead to inaccurate interpretation of data for values close to the midpoint.

**Comment by elizabethmcd** [comment]
@mezarque I think the .49 to .51 gets us closer:
Screen Shot 2024-08-13 at 3 28 53 PM

**Review by mezarque — APPROVED** [review]
Looks good to me! Thanks for addressing this.
